joern-cli/src/main/scala/io/joern/joerncli/JoernEmbed.scala

```diff
@@ -0,0 +1,144 @@
+package io.joern.joerncli
+
+import io.joern.joerncli.CpgBasedTool.exitIfInvalid
+import io.shiftleft.codepropertygraph.Cpg
+import io.shiftleft.codepropertygraph.generated.nodes.{AstNode, Method}
+
+import scala.util.Using
+import io.shiftleft.semanticcpg.language._
+import org.json4s.DefaultFormats
+import org.json4s.native.Serialization
+import overflowdb.traversal._
+
+import scala.collection.mutable
+import scala.util.hashing.MurmurHash3
+
+class BagOfAPISymbolsForMethods extends EmbeddingGenerator[Method, AstNode] {
+  override def extractObjects(cpg: Cpg): Traversal[Method]           = cpg.method
+  override def enumerateSubStructures(method: Method): List[AstNode] = method.ast.l
+  override def structureToString(astNode: AstNode): String           = astNode.code
+  override def objectToString(method: Method): String                = method.fullName
+}
+
+object EmbeddingGenerator {
+  type SparseVectorWithExplicitFeature = Map[Int, (Double, String)]
+  type SparseVector                    = Map[Int, Double]
+}
+
+/** Creates an embedding from a code property graph by following three steps: (1) Objects are extracted from the graph,
+  * each of which is ultimately to be mapped to one vector (2) For each object, enumerate its sub structures. (3) Employ
+  * feature hashing to associate each sub structure with a dimension. See "Pattern-based Vulnerability Discovery -
+  * Chapter 3"
+  */
+trait EmbeddingGenerator[T, S] {
+  import EmbeddingGenerator._
+
+  case class Embedding(data: (() => Traversal[(T, SparseVectorWithExplicitFeature)])) {
+    lazy val dimToStructure: Map[Int, String] = {
+      val m = mutable.HashMap[Int, String]()
+      data().foreach { case (_, vector) =>
+        vector.foreach { case (hash, (_, structureAsString)) =>
+          m.put(hash, structureAsString)
+        }
+      }
+      m.toMap
```

Comment:
```suggestion
      val tuples = for {
        (_, vector) <- data()
        (hash, (_, structureAsString)) <- vector
      } yield hash -> structureAsString
      
      tuples.toMap
```

Reply:
In this particular use-case, the tuples are `(dimension, structure)` pairs, and we know that for a given dimension, it is always associated with the same structure and vice versa. The reason I used this mutable map here is that I want to ensure that we don't keep all of these duplicates in memory. Otherwise, I would have preferred your version.

Reply:
👍🏻